// tsparticles/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Loader_1 = require("./Classes/Loader");
var ShapeUtils_1 = require("./Classes/Utils/ShapeUtils");
var ShapeType_1 = require("./Enums/ShapeType");
var LineDrawer_1 = require("./Classes/ShapeDrawers/LineDrawer");
var CircleDrawer_1 = require("./Classes/ShapeDrawers/CircleDrawer");
var SquareDrawer_1 = require("./Classes/ShapeDrawers/SquareDrawer");
var TriangleDrawer_1 = require("./Classes/ShapeDrawers/TriangleDrawer");
var StarDrawer_1 = require("./Classes/ShapeDrawers/StarDrawer");
var PolygonDrawer_1 = require("./Classes/ShapeDrawers/PolygonDrawer");
var TextDrawer_1 = require("./Classes/ShapeDrawers/TextDrawer");
var ImageDrawer_1 = require("./Classes/ShapeDrawers/ImageDrawer");
var Presets_1 = require("./Classes/Utils/Presets");
var Main = (function () {
    function Main() {
        this.initialized = false;
        if (typeof window !== "undefined" && window) {
            window.customRequestAnimationFrame = (function () {
                return window.requestAnimationFrame ||
                    window.webkitRequestAnimationFrame ||
                    window.mozRequestAnimationFrame ||
                    window.oRequestAnimationFrame ||
                    window.msRequestAnimationFrame ||
                    (function (callback) { return window.setTimeout(callback, 1000 / 60); });
            })();
            window.customCancelRequestAnimationFrame = (function () {
                return window.cancelAnimationFrame ||
                    window.webkitCancelRequestAnimationFrame ||
                    window.mozCancelRequestAnimationFrame ||
                    window.oCancelRequestAnimationFrame ||
                    window.msCancelRequestAnimationFrame ||
                    clearTimeout;
            })();
        }
        var squareDrawer = new SquareDrawer_1.SquareDrawer();
        var textDrawer = new TextDrawer_1.TextDrawer();
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.line, new LineDrawer_1.LineDrawer());
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.circle, new CircleDrawer_1.CircleDrawer());
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.edge, squareDrawer);
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.square, squareDrawer);
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.triangle, new TriangleDrawer_1.TriangleDrawer());
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.star, new StarDrawer_1.StarDrawer());
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.polygon, new PolygonDrawer_1.PolygonDrawer());
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.char, textDrawer);
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.character, textDrawer);
        ShapeUtils_1.ShapeUtils.addShapeDrawer(ShapeType_1.ShapeType.image, new ImageDrawer_1.ImageDrawer());
    }
    Main.prototype.init = function () {
        if (!this.initialized) {
            this.initialized = true;
            if (typeof window !== "undefined" && window) {
                var tsParticles_1 = this;
                window.particlesJS = function (tagId, params) {
                    tsParticles_1.load(tagId, params);
                };
                window.particlesJS.load = function (tagId, pathConfigJson, callback) {
                    tsParticles_1.loadJSON(tagId, pathConfigJson).then(function (container) {
                        if (container) {
                            callback(container);
                        }
                    });
                };
                window.particlesJS.setOnClickHandler = function (callback) {
                    tsParticles_1.setOnClickHandler(callback);
                };
                window.pJSDom = function () {
                    return window.tsParticles.dom();
                };
            }
        }
    };
    Main.prototype.loadFromArray = function (tagId, params, index) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, Loader_1.Loader.loadFromArray(tagId, params, index)];
            });
        });
    };
    Main.prototype.load = function (tagId, params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, Loader_1.Loader.load(tagId, params)];
            });
        });
    };
    Main.prototype.loadJSON = function (tagId, pathConfigJson) {
        return Loader_1.Loader.loadJSON(tagId, pathConfigJson);
    };
    Main.prototype.setOnClickHandler = function (callback) {
        Loader_1.Loader.setOnClickHandler(callback);
    };
    Main.prototype.dom = function () {
        return Loader_1.Loader.dom();
    };
    Main.prototype.domItem = function (index) {
        return Loader_1.Loader.domItem(index);
    };
    Main.prototype.addShape = function (shape, drawer) {
        var customDrawer;
        if (typeof drawer === "function") {
            customDrawer = {
                draw: drawer,
            };
        }
        else {
            customDrawer = drawer;
        }
        ShapeUtils_1.ShapeUtils.addShapeDrawer(shape, customDrawer);
    };
    Main.prototype.addPreset = function (preset, options) {
        Presets_1.Presets.addPreset(preset, options);
    };
    return Main;
}());
var tsParticles = new Main();
exports.tsParticles = tsParticles;
tsParticles.init();
